### PR TITLE
Generalize "keyval" data to be any custom JSON containing public portion of key

### PR DESCRIPTION
**Pull request by asraa**
Admittedly, all keys currently just contain a `public` field. This customizes it for key types (e.g. Fulcio) that may have other information in their public field (issuer, email).

THIS IS NOT THE BEST WAY TO DO IT! I'm considering an overhaul of the key types. Would love suggestions if you have a clear vision for it

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

**Comment by asraa** [comment]
@joshuagl @mnm678 the goal here is to let you implement a custom JSON for fulcio signers (the big motivation)

**Review by dlorenc — APPROVED** [review]
LGTM!

File: data/types.go [hunk 1]
```diff
@@ -33,15 +33,42 @@ type Signature struct {
...
+	Type       string          `json:"keytype"`
+	Scheme     string          `json:"scheme"`
+	Algorithms []string        `json:"keyid_hash_algorithms,omitempty"`
+	Value      json.RawMessage `json:"keyval"`
 
 	ids    []string
 	idOnce sync.Once
 }
 
+// A KeyType instance is an implementation of a KeyType
```
Comment (trishankatdatadog) [inline data/types.go:45]: This comment doesn't make sense...

**Comment by trishankatdatadog** [comment]
Thanks for your PR, Asra!

Do we have a corresponding PR to the TUF spec suggesting this generalization of the public key value?

**Comment by asraa** [comment]
It's compliant with TUF spec as far as I read, since the specification defines keyval as a dictionary containing the public portion of the key, but does not define a general format for each keytype. https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#file-formats-general-principles--file-formats-general-principles

>  : <dfn>KEYVAL</dfn>
  ::
    A dictionary containing the public portion of the key.

Once oidc keytypes are implemented, then the spec would need to be updated to include the keyval dictionary with the email and issuer for those types.

File: data/types.go (hunk 1 above)
Comment (asraa) [reply data/types.go:45]: fixed

**Comment by trishankatdatadog** [comment]
> It's compliant with TUF spec as far as I read, since the specification defines keyval as a dictionary containing the public portion of the key, but does not define a general format for each keytype. https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#file-formats-general-principles--file-formats-general-principles
> 
> > : KEYVAL
> > ::
> > A dictionary containing the public portion of the key.
>

You are right: https://theupdateframework.github.io/specification/v1.0.20/index.html#keyval
 
> Once oidc keytypes are implemented, then the spec would need to be updated to include the keyval dictionary with the email and issuer for those types.

Yes, it would be another keytype.

Now, does your PR allow keyvals to remain strings, not just arbitrary JSON objects?

**Comment by trishankatdatadog** [comment]
> Now, does your PR allow keyvals to remain strings, not just arbitrary JSON objects?

@asraa: as long as go-tuf can handle _both_ strings and JSON now for PUBLIC, I am happy to approve ✅

**Comment by asraa** [comment]
> Now, does your PR allow keyvals to remain strings, not just arbitrary JSON objects?

Ah! I didn't totally understand the question, but yes, it preserves existing behavior! Since 
{ "public": "string" } is a potential json.RawMessage to embed inside `KeyVal`

**Comment by trishankatdatadog** [comment]
> Ah! I didn't totally understand the question, but yes, it preserves existing behavior! Since
> { "public": "string" } is a potential json.RawMessage to embed inside `KeyVal`

💯 

Spiritually approved because: (1) it doesn't violate the spec, and (2) existing behaviour is preserved.

Let me just sift through the code for implementation details, and then we are good to go!

**Comment by asraa** [comment]
superseded by https://github.com/theupdateframework/go-tuf/pull/148

File: data/types.go [hunk 2]
```diff
@@ -33,15 +33,42 @@ type Signature struct {
...
+type ed25519KeyType struct{}
+type p256KeyType struct{}
+
+func (v ed25519KeyType) UniquePublic(valueBytes json.RawMessage) string {
+	s := KeyValue{}
+	json.Unmarshal(valueBytes, &s)
+	return s.Public.String()
+}
+
+func (v p256KeyType) UniquePublic(valueBytes json.RawMessage) string {
```
[the remark this answers is not in the document]
Comment (asraa) [reply data/types.go:66]: yes! actually now it is a method that each key type implements in https://github.com/theupdateframework/go-tuf/pull/148

**Pull request closed without merging**